engine/execution/ingestion/finalized_and_executed_handler.go

```diff
@@ -0,0 +1,158 @@
+package ingestion
+
+import (
+	"context"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type OnBlockFinalizedAndExecutedConsumer = func(h *flow.Header)
+
+// FinalizedAndExecutedDispatcher dispatches on block finalized and executed events
+type FinalizedAndExecutedDispatcher interface {
+	AddOnFinalizedAndExecutedConsumer(OnBlockFinalizedAndExecutedConsumer)
+	OnBlockExecuted(ctx context.Context, h *flow.Header) error
+	OnBlockFinalized(ctx context.Context, h *flow.Header) error
+}
+
+type finalizedAndExecutedDispatcher struct {
+	sync.RWMutex
+	log zerolog.Logger
+
+	blockFinalizedAndExecutedConsumers []OnBlockFinalizedAndExecutedConsumer
+	latestFinalizedAndExecutedHeight   uint64
+
+	isBlockFinalized func(context.Context, *flow.Header) (bool, error)
+	isBlockExecuted  func(context.Context, *flow.Header) (bool, error)
+}
+
+type FinalizedAndExecutedDispatcherOption func(*finalizedAndExecutedDispatcher)
+
+// FinalizedAndExecutedDispatcherWithLogger sets the logger for the
+// FinalizedAndExecutedDispatcher and adds a component field to the logger
+func FinalizedAndExecutedDispatcherWithLogger(
+	log zerolog.Logger,
+) FinalizedAndExecutedDispatcherOption {
+	return func(f *finalizedAndExecutedDispatcher) {
+		f.log = log.With().
+			Str("component", "finalized-and-executed-dispatcher").
+			Logger()
+	}
+}
+
+// NewFinalizedAndExecutedDispatcher returns a new FinalizedAndExecutedDispatcher
+func NewFinalizedAndExecutedDispatcher(
+	latestFinalizedAndExecutedHeight uint64,
+	isBlockFinalized func(context.Context, *flow.Header) (bool, error),
+	isBlockExecuted func(context.Context, *flow.Header) (bool, error),
+	options ...FinalizedAndExecutedDispatcherOption,
+) FinalizedAndExecutedDispatcher {
+	d := &finalizedAndExecutedDispatcher{
+		log: zerolog.Nop(),
+
+		blockFinalizedAndExecutedConsumers: []OnBlockFinalizedAndExecutedConsumer{},
+		latestFinalizedAndExecutedHeight:   latestFinalizedAndExecutedHeight,
+
+		isBlockFinalized: isBlockFinalized,
+		isBlockExecuted:  isBlockExecuted,
+	}
+
+	for _, option := range options {
+		option(d)
+	}
+
+	return d
+}
+
+func (f *finalizedAndExecutedDispatcher) AddOnFinalizedAndExecutedConsumer(
+	consumer OnBlockFinalizedAndExecutedConsumer,
+) {
+	f.Lock()
+	defer f.Unlock()
+
+	f.blockFinalizedAndExecutedConsumers =
+		append(f.blockFinalizedAndExecutedConsumers, consumer)
+}
+
+func (f *finalizedAndExecutedDispatcher) OnBlockExecuted(
+	ctx context.Context,
+	h *flow.Header,
+) error {
+	isNew, err := f.checkOnBlockExecuted(ctx, h)
```

Comment:
We can skip this check if `h.Height != latestFinalizedAndExecutedHeight + 1`.

Likewise, we can skip the same check for OnBlockFinalized

Reply:
good point. I will change this.